lua-api-crates/table-funcs/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+            let (key, value) = pair?;
+            tbl_vec.push((key, value));
+        }
+    }
+    let tbl_len = tbl_vec.len();
+    // note we might allocate a bit too much here, but in many use cases we will be correct
+    let tbl: Table<'lua> = lua.create_table_with_capacity(0, tbl_len)?;
+
+    let behavior = behavior.unwrap_or_default();
+    for (key, value) in tbl_vec {
+        if !tbl.contains_key(key.clone())? {
+            tbl.set(key, value)?;
+        } else if let LuaValue::Table(t) = value {
+            let inner_tbl = deep_extend(lua, (vec![tbl.get(key.clone())?, t], Some(behavior)))?;
+            tbl.set(key, inner_tbl)?;
+        } else if behavior == ConflictMode::Force {
+            tbl.set(key, value)?;
+        } else if behavior == ConflictMode::Error {
+            return Err(mlua::Error::runtime(format!(
+                "The key {} is in more than one of the tables.",
+                key.to_string()?
+            )));
+        }
+    }
+
+    Ok(tbl)
+}
+
+fn clone<'lua>(
+    lua: &'lua Lua,
+    (table, behavior): (Table<'lua>, Option<DepthMode>),
+) -> mlua::Result<Table<'lua>> {
+    let res: Table<'lua> = lua.create_table()?;
+
+    let behavior = behavior.unwrap_or_default();
+    for pair in table.pairs::<LuaValue, LuaValue>() {
+        let (key, value) = pair?;
+        match behavior {
+            DepthMode::Top => res.set(key, value)?,
+            DepthMode::Deep => {
+                if let LuaValue::Table(tbl) = value {
+                    res.set(key, clone(lua, (tbl, Some(behavior)))?)?
+                } else {
+                    res.set(key, value)?;
+                }
+            }
+        }
+    }
+    Ok(res)
+}
+
+fn flatten<'lua>(
+    lua: &'lua Lua,
+    (arrays, behavior): (Vec<LuaValue<'lua>>, Option<DepthMode>),
+) -> mlua::Result<Vec<LuaValue<'lua>>> {
+    let mut flat_vec: Vec<LuaValue> = vec![];
+    let behavior = behavior.unwrap_or_default();
+    for item in arrays {
+        match item {
+            LuaValue::Table(tbl) => {
+                if behavior == DepthMode::Deep {
+                    let tbl_as_vec = tbl.sequence_values().filter_map(|x| x.ok()).collect();
+                    let mut flat = flatten(lua, (tbl_as_vec, Some(behavior)))?;
+                    flat_vec.append(&mut flat);
+                } else {
+                    for elem in tbl.sequence_values::<LuaValue>() {
+                        flat_vec.push(elem?);
+                    }
+                }
+            }
+            LuaValue::Nil => (),
+            other => {
+                flat_vec.push(other);
+            }
+        }
+    }
+    Ok(flat_vec)
+}
+
+/// note that the `#` operator only works correctly on arrays in Lua
+fn count<'lua>(_: &'lua Lua, table: Table<'lua>) -> mlua::Result<usize> {
+    Ok(table.pairs::<LuaValue, LuaValue>().count())
+}
+
+fn get<'lua>(
+    _: &'lua Lua,
+    (table, key, mut extra_keys): (Table<'lua>, LuaValue<'lua>, LuaMultiValue<'lua>),
```

Comment:
FWIW, I think this whole function can be simplified to avoid the special case of the initial key, something like this:

```rust
fn get<'lua>(
    _: &'lua Lua,
    (mut value, mut keys): (Table<'lua>, LuaMultiValue<'lua>)
) -> mlua::Result<LuaValue<'lua>> {
     for key in keys {
        match value {
             LuaValue::Table(tbl) => {
                  value = tbl.get(key)?;
             }
             _ => {
               // Cannot index non-table
               return Ok(LuaValue::Nil)
            }
         }
     }
     Ok(value)
}
```

Reply:
> FWIW, I think this whole function can be simplified to avoid the special case of the initial key, something like this:
> 
> ```rust
> fn get<'lua>(
>     _: &'lua Lua,
>     (mut value, mut keys): (Table<'lua>, LuaMultiValue<'lua>)
> ) -> mlua::Result<LuaValue<'lua>> {
>      for key in keys {
>         match value {
>              LuaValue::Table(tbl) => {
>                   value = tbl.get(key)?;
>              }
>              _ => {
>                // Cannot index non-table
>                return Ok(LuaValue::Nil)
>             }
>          }
>      }
>      Ok(value)
> }
> ```

I did consider that too, but choose the other way to hopefully have better error messages. If you prefer this way, I am fine with changing it.

Reply:
I did something similar to what was mentioned here now, but I used `.get` from `luahelper` on the `UserData` instead of the above.